core/mvcc/database/mod.rs

```diff
@@ -1753,40 +1844,37 @@ impl<Clock: LogicalClock> MvStore<Clock> {
         last_rowid
     }
 
-    pub fn needs_recover(&self) -> bool {
-        self.storage.needs_recover()
-    }
-
-    pub fn mark_recovered(&self) {
-        self.storage.mark_recovered();
-    }
-
     pub fn get_logical_log_file(&self) -> Arc<dyn File> {
         self.storage.get_logical_log_file()
     }
 
-    pub fn recover_logical_log(&self, io: &Arc<dyn crate::IO>, pager: &Arc<Pager>) -> Result<()> {
-        // Get lock, if we don't get it we will wait until recover finishes in another connection
-        // and then return.
-        let _recover_guard = self.recover_lock.write();
-        if !self.storage.needs_recover() {
-            // another connection completed recover
-            return Ok(());
-        }
+    // Recovers the logical log if there is any content.
+    // Returns true if the logical log was recovered, false otherwise.
+    pub fn maybe_recover_logical_log(&self, pager: Arc<Pager>) -> Result<bool> {
         let file = self.get_logical_log_file();
         let mut reader = StreamingLogicalLogReader::new(file.clone());
 
+        if file.size()? == 0 {
+            return Ok(false);
+        }
+
         let c = reader.read_header()?;
-        io.wait_for_completion(c)?;
+        pager.io.wait_for_completion(c)?;
         let tx_id = 0;
         self.begin_load_tx(pager.clone())?;
         loop {
-            match reader.next_record(io).unwrap() {
+            match reader.next_record(&pager.io).unwrap() {
                 StreamingResult::InsertRow { row, rowid } => {
-                    tracing::trace!("read {rowid:?}");
+                    tracing::trace!("read {row:?} with rowid {rowid:?}");
+                    if self.table_id_to_rootpage.get(&row.id.table_id).is_none() {
+                        self.insert_table_id_to_rootpage(row.id.table_id, None);
```

Comment:
In this case, what about checking the row is an insert into `sqlite_schema` and then call `insert_table_id_to_rootpage`. If table_id is not found and insert is not a create table like row, it should probably panic

Reply:
done